### PR TITLE
Issue-4514 | Add configurable annotation to pvc template

### PR DESCRIPTION
**Pull request by santoshr1016**
## Description

The works adds the default annotations field to the PVC templates

## Related Issues
https://github.com/astronomer/issues/issues/4514

## Testing

Unit tests are included. There were changes to ES, stan, registry, prometheus, and alertmanager, so it would be good for QA to add some arbitrary annotations when they are testing and make sure everything is fine.

## Merging

This should go into 0.29. Possibly 0.28.

### CODE REVIEW

File: charts/elasticsearch/values.yaml [hunk 1]
```diff
@@ -50,6 +50,10 @@ common:
   # associated persistent volume claims.
   persistence:
     enabled: true
+    annotations: {
```
Comment (pgvishnuram) [inline charts/elasticsearch/values.yaml:53]: @santoshr1016 - hope these are added by mistake in here

File: charts/alertmanager/values.yaml [hunk 2]
```diff
@@ -30,6 +30,10 @@ persistence:
   enabled: true
   size: 2Gi
   storageClassName: ~
+  annotations: {
```
Comment (pgvishnuram) [inline charts/alertmanager/values.yaml:33]: @santoshr1016 - hope these are added by mistake in here

Comment (pgvishnuram) [reply charts/alertmanager/values.yaml:33]: by default its annotations: {}

File: tests/chart_tests/test_sts_volume_claim_annotations.py [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+                "charts/astronomer/templates/registry/registry-statefulset.yaml",
+                "charts/prometheus/templates/prometheus-statefulset.yaml",
+                "charts/stan/templates/statefulset.yaml",
+            ],
+        )
+
+        assert len(docs) == 6
+        for doc in docs:
+            if "annotations" in doc["spec"]["volumeClaimTemplates"][0]["metadata"]:
+                assert doc["spec"]["volumeClaimTemplates"][0]["metadata"]["annotations"]
```
Comment (danielhoherd) [inline tests/chart_tests/test_sts_volume_claim_annotations.py:30]: We should set `.Values.persistence.annotations` on line 16 and make some assertions about them here

Comment (pgvishnuram) [reply tests/chart_tests/test_sts_volume_claim_annotations.py:30]: this case if for default and we need one more test case for overrides with custom annotations added

Comment (danielhoherd) [reply tests/chart_tests/test_sts_volume_claim_annotations.py:30]: If annotations are there, we should assert something. If they are not there, we should assert that they are not there.

File: tests/chart_tests/test_sts_volume_claim_annotations.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+            kube_version=kube_version,
+            values={},
+            show_only=[
+                "charts/elasticsearch/templates/master/es-master-statefulset.yaml",
+                "charts/elasticsearch/templates/data/es-data-statefulset.yaml",
+                "charts/alertmanager/templates/alertmanager-statefulset.yaml",
+                "charts/astronomer/templates/registry/registry-statefulset.yaml",
+                "charts/prometheus/templates/prometheus-statefulset.yaml",
+                "charts/stan/templates/statefulset.yaml",
+            ],
```
Comment (danielhoherd) [inline tests/chart_tests/test_sts_volume_claim_annotations.py:24]: Can we replace this with something that auto-discovers sts? We could either build this list by walking the filesystem, or we could render the full chart and filter it for sts, which is how we've done this elsewhere. <https://github.com/astronomer/astronomer/blob/d46aacf2ec/tests/chart_tests/test_default_chart.py#L10-L18>

File: tests/chart_tests/test_sts_volume_claim_annotations.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+        )
+
+        assert len(docs) == 6
+
+        with pytest.raises(Exception) as exec_info:
+            for doc in docs:
+                if doc["spec"]["volumeClaimTemplates"][0]["metadata"]["annotations"]:
+                    raise Exception("Empty Annotations")
+            assert exec_info.value.args[0] == "Empty Annotations"
+            assert str(exec_info.value) == "Empty Annotations"
```
Comment (danielhoherd) [inline tests/chart_tests/test_sts_volume_claim_annotations.py:34]: Do we need this `pytest.raises` context manager? Can we `assert not doc["spec"]["volumeClaimTemplates"][0]["metadata"]["annotations"]` or something with simpler syntax?

File: tests/chart_tests/test_sts_volume_claim_annotations.py (hunk 4 above)
Comment (danielhoherd) [reply tests/chart_tests/test_sts_volume_claim_annotations.py:24]: OR change the docstring to not say "Test **all** sts"

File: tests/chart_tests/test_sts_volume_claim_annotations.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+        )
+        assert len(docs) == 1
+        for doc in docs:
+            assert {
+                "annotation": "astro-test",
+                "astro.io/monitoring": "prom-sts",
+            } == doc["spec"]["volumeClaimTemplates"][0]["metadata"]["annotations"]
+
+    def test_registry_sts_with_overridden_annotations(self, kube_version):
+        """Test all sts for the volume claim templates annotations"""
```
Comment (danielhoherd) [inline tests/chart_tests/test_sts_volume_claim_annotations.py:89]: Need to change this wording too.

File: tests/chart_tests/test_sts_volume_claim_annotations.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+                    }
+                }
+            },
+            show_only=[
+                "charts/stan/templates/statefulset.yaml",
+            ],
+        )
+        assert len(docs) == 1
+        for doc in docs:
+            # print(doc["spec"]["volumeClaimTemplates"][0]["metadata"]["annotations"])
```
Comment (danielhoherd) [inline tests/chart_tests/test_sts_volume_claim_annotations.py:135]:
```suggestion
```

File: tests/chart_tests/test_sts_volume_claim_annotations.py [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        """Test if the sts contains volume claim templates annotations or not"""
+        sts_workload = [doc for doc in render_chart() if doc["kind"] == "StatefulSet"]
+        assert len(sts_workload) == 7
+
+        with pytest.raises(Exception) as except_info:
+            for doc in sts_workload:
+                if doc["spec"]["volumeClaimTemplates"][0]["metadata"]["annotations"]:
+                    raise Exception("Empty Annotations")
+            assert except_info.value.args[0] == "Empty Annotations"
+            assert str(except_info.value) == "Empty Annotations"
```
Comment (danielhoherd) [inline tests/chart_tests/test_sts_volume_claim_annotations.py:22]: We can do this without pytest.raises

```suggestion
        vcts = [
            doc["spec"]["volumeClaimTemplates"]
            for doc in render_chart()
            if doc["kind"] == "StatefulSet" and doc["spec"].get("volumeClaimTemplates")
        ]
        assert len(vcts) == 6

        for vct in vcts:
            assert not vct[0]["metadata"].get("annotations")
```

**Review by danielhoherd — APPROVED** [review]
```
 _____       ______  _________  ____    ____
|_   _|    .' ___  ||  _   _  ||_   \  /   _|
  | |     / .'   \_||_/ | | \_|  |   \/   |
  | |   _ | |   ____    | |      | |\  /| |
 _| |__/ |\ `.___]  |  _| |_    _| |_\/_| |_
|________| `._____.'  |_____|  |_____||_____|
```

**Comment by danielhoherd** [comment]
@santoshr1016 can you fill out the PR description here?

**Pull request merged**